File: API/SafeDesk365.SDK/Api/DeskAvailabilities/DeskAvailabilitiesRequestBuilder.cs
using Microsoft.Kiota.Abstractions;
using Microsoft.Kiota.Abstractions.Serialization;
using SafeDesk365.SDK.Api.DeskAvailabilities.Date;
using SafeDesk365.SDK.Api.DeskAvailabilities.Location;
using SafeDesk365.SDK.Api.DeskAvailabilities.Upcoming;
using SafeDesk365.SDK.Models;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
namespace SafeDesk365.SDK.Api.DeskAvailabilities {
    /// <summary>Builds and executes requests for operations under \api\deskAvailabilities</summary>
    public class DeskAvailabilitiesRequestBuilder {
        /// <summary>The date property</summary>
        public DateRequestBuilder Date { get =>
            new DateRequestBuilder(PathParameters, RequestAdapter);
        }
        /// <summary>The location property</summary>
        public LocationRequestBuilder Location { get =>
            new LocationRequestBuilder(PathParameters, RequestAdapter);
        }
        /// <summary>Path parameters for the request</summary>
        private Dictionary<string, object> PathParameters { get; set; }
        /// <summary>The request adapter to use to execute the requests.</summary>
        private IRequestAdapter RequestAdapter { get; set; }
        /// <summary>The upcoming property</summary>
        public UpcomingRequestBuilder Upcoming { get =>
            new UpcomingRequestBuilder(PathParameters, RequestAdapter);
        }
        /// <summary>Url template to use to build the URL for the current request builder</summary>
        private string UrlTemplate { get; set; }
        /// <summary>
        /// Instantiates a new DeskAvailabilitiesRequestBuilder and sets the default values.
        /// <param name="pathParameters">Path parameters for the request</param>
        /// <param name="requestAdapter">The request adapter to use to execute the requests.</param>
        /// </summary>
        public DeskAvailabilitiesRequestBuilder(Dictionary<string, object> pathParameters, IRequestAdapter requestAdapter) {
            _ = pathParameters ?? throw new ArgumentNullException(nameof(pathParameters));
            _ = requestAdapter ?? throw new ArgumentNullException(nameof(requestAdapter));
            UrlTemplate = "{+baseurl}/api/deskAvailabilities";
            var urlTplParams = new Dictionary<string, object>(pathParameters);
            PathParameters = urlTplParams;
            RequestAdapter = requestAdapter;
        }
        /// <summary>
        /// Instantiates a new DeskAvailabilitiesRequestBuilder and sets the default values.
        /// <param name="rawUrl">The raw URL to use for the request builder.</param>
        /// <param name="requestAdapter">The request adapter to use to execute the requests.</param>
        /// </summary>
        public DeskAvailabilitiesRequestBuilder(string rawUrl, IRequestAdapter requestAdapter) {
            if(string.IsNullOrEmpty(rawUrl)) throw new ArgumentNullException(nameof(rawUrl));
            _ = requestAdapter ?? throw new ArgumentNullException(nameof(requestAdapter));
            UrlTemplate = "{+baseurl}/api/deskAvailabilities";
            var urlTplParams = new Dictionary<string, object>();
            urlTplParams.Add("request-raw-url", rawUrl);
            PathParameters = urlTplParams;
            RequestAdapter = requestAdapter;
        }
        public RequestInformation CreateGetRequestInformation(Action<IDictionary<string, string>> headers = default, IEnumerable<IRequestOption> options = default) {
            var requestInfo = new RequestInformation {
                HttpMethod = Method.GET,
                UrlTemplate = UrlTemplate,
                PathParameters = PathParameters,
            };
            headers?.Invoke(requestInfo.Headers);
            requestInfo.AddRequestOptions(options?.ToArray());
            return requestInfo;
        }
        public async Task<IEnumerable<DeskAvailability>> GetAsync(Action<IDictionary<string, string>> headers = default, IEnumerable<IRequestOption> options = default, IResponseHandler responseHandler = default, CancellationToken cancellationToken = default) {
            var requestInfo = CreateGetRequestInformation(headers, options);
            return await RequestAdapter.SendCollectionAsync<DeskAvailability>(requestInfo, DeskAvailability.CreateFromDiscriminatorValue, responseHandler, default, cancellationToken);
        }
    }
}

File: SDK/NPM/api/deskAvailabilities/deskAvailabilitiesRequestBuilder.ts
import {DeskAvailability} from '../../models/';
import {createDeskAvailabilityFromDiscriminatorValue} from '../../models/createDeskAvailabilityFromDiscriminatorValue';
import {DateRequestBuilder} from './date/dateRequestBuilder';
import {WithDateItemRequestBuilder} from './date/item/withDateItemRequestBuilder';
import {WithLocationItemRequestBuilder} from './location/item/withLocationItemRequestBuilder';
import {LocationRequestBuilder} from './location/locationRequestBuilder';
import {UpcomingRequestBuilder} from './upcoming/upcomingRequestBuilder';
import {getPathParameters, HttpMethod, Parsable, ParsableFactory, RequestAdapter, RequestInformation, RequestOption, ResponseHandler} from '@microsoft/kiota-abstractions';

/** Builds and executes requests for operations under /api/deskAvailabilities  */
export class DeskAvailabilitiesRequestBuilder {
    /** The date property  */
    public get date(): DateRequestBuilder {
        return new DateRequestBuilder(this.pathParameters, this.requestAdapter);
    }
    /** The location property  */
    public get location(): LocationRequestBuilder {
        return new LocationRequestBuilder(this.pathParameters, this.requestAdapter);
    }
    /** Path parameters for the request  */
    private readonly pathParameters: Record<string, unknown>;
    /** The request adapter to use to execute the requests.  */
    private readonly requestAdapter: RequestAdapter;
    /** The upcoming property  */
    public get upcoming(): UpcomingRequestBuilder {
        return new UpcomingRequestBuilder(this.pathParameters, this.requestAdapter);
    }
    /** Url template to use to build the URL for the current request builder  */
    private readonly urlTemplate: string;
    /**
     * Instantiates a new DeskAvailabilitiesRequestBuilder and sets the default values.
     * @param pathParameters The raw url or the Url template parameters for the request.
     * @param requestAdapter The request adapter to use to execute the requests.
     */
    public constructor(pathParameters: Record<string, unknown> | string | undefined, requestAdapter: RequestAdapter) {
        if(!pathParameters) throw new Error("pathParameters cannot be undefined");
        if(!requestAdapter) throw new Error("requestAdapter cannot be undefined");
        this.urlTemplate = "{+baseurl}/api/deskAvailabilities";
        const urlTplParams = getPathParameters(pathParameters);
        this.pathParameters = urlTplParams;
        this.requestAdapter = requestAdapter;
    };
    public createGetRequestInformation(headers?: Record<string, string> | undefined, options?: RequestOption[] | undefined) : RequestInformation {
        const requestInfo = new RequestInformation();
        requestInfo.urlTemplate = this.urlTemplate;
        requestInfo.pathParameters = this.pathParameters;
        requestInfo.httpMethod = HttpMethod.GET;
        if(headers) requestInfo.headers = headers;
        options && requestInfo.addRequestOptions(...options);
        return requestInfo;
    };
    /**
     * Gets an item from the SafeDesk.API.api.deskAvailabilities.date.item collection
     * @param id Unique identifier of the item
     * @returns a WithDateItemRequestBuilder
     */
    public dateById(id: string) : WithDateItemRequestBuilder {
        if(!id) throw new Error("id cannot be undefined");
        const urlTplParams = getPathParameters(this.pathParameters);
        urlTplParams["date"] = id
        return new WithDateItemRequestBuilder(urlTplParams, this.requestAdapter);
    };
    public get(headers?: Record<string, string> | undefined, options?: RequestOption[] | undefined, responseHandler?: ResponseHandler | undefined) : Promise<DeskAvailability[] | undefined> {
        const requestInfo = this.createGetRequestInformation(
            headers, options
        );
        return this.requestAdapter?.sendCollectionAsync<DeskAvailability>(requestInfo, createDeskAvailabilityFromDiscriminatorValue, responseHandler, undefined) ?? Promise.reject(new Error('http core is null'));
    };
    /**
     * Gets an item from the SafeDesk.API.api.deskAvailabilities.location.item collection
     * @param id Unique identifier of the item
     * @returns a WithLocationItemRequestBuilder
     */
    public locationById(id: string) : WithLocationItemRequestBuilder {
        if(!id) throw new Error("id cannot be undefined");
        const urlTplParams = getPathParameters(this.pathParameters);
        urlTplParams["location"] = id
        return new WithLocationItemRequestBuilder(urlTplParams, this.requestAdapter);
    };
}
